community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java

```diff
@@ -19,72 +19,137 @@
     {
-        this.array = arrayFactory.newDynamicByteArray( chunkSize, DEFAULT_VALUE );
+        this.chunkSize = chunkSize;
+        this.array = arrayFactory.newDynamicByteArray( chunkSize, minusOneBytes( ID_AND_COUNT_SIZE ) );
         this.denseNodeThreshold = denseNodeThreshold;
         this.relGroupCache = new RelGroupCache( arrayFactory, chunkSize, base );
     }
 
+    private static byte[] minusOneBytes( int length )
+    {
+        byte[] bytes = new byte[length];
+        Arrays.fill( bytes, (byte) -1 );
+        return bytes;
+    }
+
     /**
      * Increment relationship count for {@code nodeId}.
      * @param nodeId node to increment relationship count for.
      * @return count after the increment.
      */
     public int incrementCount( long nodeId )
     {
-        ByteArray array = this.array.at( nodeId );
-        int count = getCount( array, nodeId ) + 1;
-        setCount( array, nodeId, count );
-        return count;
+        return incrementCount( array, nodeId, SPARSE_COUNT_OFFSET );
     }
 
-    private void setCount( ByteArray array, long nodeId, int count )
+    void setCount( long nodeId, int count )
     {
+        assertValidCount( nodeId, count );
         array.setInt( nodeId, SPARSE_COUNT_OFFSET, count );
     }
 
-    private static int getCount( ByteArray array, long nodeId )
+    private static void assertValidCount( long nodeId, int count )
     {
-        return array.getInt( nodeId, SPARSE_COUNT_OFFSET );
+        if ( count > MAX_COUNT )
+        {
+            // Meaning there are bits outside of this mask, meaning this value is too big
+            throw new IllegalStateException( "Tried to increment count of " + nodeId + " to " + count +
+                    ", which is too big in one single import" );
+        }
     }
 
+    /**
+     * Called by the one calling {@link #incrementCount(long)} after all nodes have been added.
+     * Done like this since currently it's just overhead trying to maintain a high id in the face
+     * of current updates, whereas it's much simpler to do this from the code incrementing the counts.
+     *
+     * @param nodeId high node id in the store, e.g. the highest node id + 1
+     */
+    public void setHighNodeId( long nodeId )
+    {
+        this.highId = nodeId;
+        this.chunkChangedArray = new byte[chunkOf( nodeId ) + 1];
+    }
+
+    private static int getCount( ByteArray array, long index, int offset )
+    {
+        long rawCount = array.getInt( index, offset ) & COUNT_MASK;
+        if ( rawCount == COUNT_MASK )
+        {
+            return 0;
+        }
+        return (int) rawCount;
+    }
+
+    private static int incrementCount( ByteArray array, long index, int offset )
+    {
+        array = array.at( index );
+        int count = getCount( array, index, offset ) + 1;
+        assertValidCount( index, count );
+        array.setInt( index, offset, count );
+        return count;
+    }
```

Comment:
It does not. There are external factors making it so that there's only a single thread at a time changing one particular index in that array. See for example `CalculateDenseNodesStage` and its steps, particularly `CalculateDenseNodePrepareStep`